v2/tools/asoctl/internal/crd/cleaner.go

```diff
@@ -0,0 +1,193 @@
+}
+
+func NewCleaner(apiExtensionsClient apiextensionsclient.CustomResourceDefinitionInterface, client client.Client, dryRun bool) *Cleaner {
+	migrationBackoff := wait.Backoff{ // TODO: Still need to see if we want exponential backoff or linear is fine
+		Duration: 2 * time.Second, // wait 2s between attempts, this will help us in a state of conflict.
+		Steps:    3,               // 3 retry on error attempts per object
+	}
+	return &Cleaner{
+		apiExtensionsClient: apiExtensionsClient,
+		client:              client,
+		migrationBackoff:    migrationBackoff,
+		dryRun:              dryRun,
+	}
+}
+
+func (c *Cleaner) Run(ctx context.Context) error {
+	list, err := c.apiExtensionsClient.List(ctx, v1.ListOptions{})
+	if err != nil {
+		return errors.Wrap(err, "failed to list CRDs")
+	}
+
+	if list == nil || len(list.Items) == 0 {
+		return errors.New("found 0 results, make sure you have ASO CRDs installed")
+
+	}
+
+	var updated int
+	crdRegexp := regexp.MustCompile(`.*\.azure\.com`)
+	deprecatedVersionRegexp := regexp.MustCompile(`v1alpha1api\d{8}(preview)?(storage)?`)
+	for _, crd := range list.Items {
+		crd := crd
+
+		if !crdRegexp.MatchString(crd.Name) {
+			continue
+		}
+
+		newStoredVersions, matchedStoredVersion := removeMatchingStoredVersions(crd.Status.StoredVersions, deprecatedVersionRegexp)
+
+		// In the check below, where CRD only has an only matched version, then the newStoredVersions == 0
+		if len(newStoredVersions) > 0 && len(newStoredVersions) != len(crd.Status.StoredVersions) {
+			klog.Infof("starting cleanup for %q", crd.Name)
+			objectsToMigrate, err := c.getObjectsForMigration(ctx, crd, matchedStoredVersion)
+			if err != nil {
+				return err
+			}
+
+			err = c.migrateObjects(ctx, objectsToMigrate)
+			if err != nil {
+				return err
+			}
+
+			err = c.updateStorageVersions(ctx, crd, newStoredVersions)
+			if err != nil {
+				return err
+			}
+
+			updated++
+		} else {
+			klog.Infof("nothing to update for %q\n", crd.Name)
+		}
+	}
+
+	if !c.dryRun {
+		klog.Infof("updated %d CRD(s)\n", updated)
+
+	}
+
+	return nil
+}
+
+func (c *Cleaner) updateStorageVersions(
+	ctx context.Context,
+	crd apiextensions.CustomResourceDefinition,
+	newStoredVersions []string) error {
+
+	if c.dryRun {
+		klog.Infof("would update storedVersions for %q CRD to: %s\n", crd.Name, newStoredVersions)
+		return nil
+	}
+
+	crd.Status.StoredVersions = newStoredVersions
+	updatedCrd, err := c.apiExtensionsClient.UpdateStatus(ctx, &crd, v1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+	klog.Infof("updated %q CRD status storedVersions to : %s\n", crd.Name, updatedCrd.Status.StoredVersions)
+
+	return nil
+}
+
+func (c *Cleaner) migrateObjects(ctx context.Context, objectsToMigrate *unstructured.UnstructuredList) error {
+	for _, obj := range objectsToMigrate.Items {
+		obj := obj
+		if c.dryRun {
+			klog.V(logging.Verbose).Infof("resource %q to migrate for kind %q", obj.GetName(), obj.GroupVersionKind().Kind)
```

Comment:
Phrasing here is a bit clunky - here's a suggestion, feel free to take a different approach:

```suggestion
			klog.V(logging.Verbose).Infof("Would migrate resource %q of kind %q", obj.GetName(), obj.GroupVersionKind().Kind)
```